## Preliminary EDA (Exploratory Data Analysis)
Imagine you are a data scientist at a real estate company. Your task is to analyze Atlanta's real estate data. Start by inspecting the dataset *real_estate_ga.csv*, which contains property information from 2021. Familiarize yourself with its structure and contents before proceeding with EDA.

**Requirements:**

- Please use [pandas.DataFrame](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html) to manipulate data.

- Please follow the python code style (https://www.python.org/dev/peps/pep-0008/). If TA finds your code hard to read, you will lose points. This requirement will stay for the whole semester.

In [1]:
import numpy as np
import pandas as pd

#### 1. First, load the data as a DataFrame.

In [2]:
df = pd.read_csv("real_estate_ga.csv")

#### 2. Display the first 4 rows of the DataFrame `df`.

In [3]:
df.head(4)

,Unnamed: 0,id,stateId,countyId,cityId,country,datePostedString,is_bankOwned,is_forAuction,event,...,parking,garageSpaces,hasGarage,levels,pool,spa,isNewConstruction,hasPetsAllowed,homeType,county
0,0,31503-110785431,16,17,55064,USA,7/12/2021,0,0,Listed for sale,...,0,0,0,0,0,0,0,0,SINGLE_FAMILY,Brantley County
1,1,31503-76611082,16,18,55064,USA,7/12/2021,0,0,Listed for sale,...,0,0,0,0,0,0,0,0,SINGLE_FAMILY,Ware County
2,2,31503-93126153,16,19,55064,USA,7/10/2021,0,0,Listed for sale,...,1,0,0,0,0,0,0,0,SINGLE_FAMILY,Ware County
3,3,31503-110785598,16,20,55064,USA,7/9/2021,0,0,Listed for sale,...,0,0,0,0,0,0,0,0,SINGLE_FAMILY,Brantley County


#### 3. Display the 3 random rows of the DataFrame `df`.

In [4]:
df.sample(3)

,Unnamed: 0,id,stateId,countyId,cityId,country,datePostedString,is_bankOwned,is_forAuction,event,...,parking,garageSpaces,hasGarage,levels,pool,spa,isNewConstruction,hasPetsAllowed,homeType,county
3054,3054,30257-69963444,16,111482,5971,USA,4/9/2021,0,0,Listed for sale,...,1,0,0,One and One Half,0,0,0,0,SINGLE_FAMILY,Lamar County
5281,5281,30224-69917071,16,183543,11738,USA,7/10/2021,0,0,Listed for sale,...,0,0,0,One,0,1,0,0,SINGLE_FAMILY,Spalding County
11483,11483,30552-2078362435,16,40415,5515,USA,6/25/2021,0,0,Listing removed,...,0,0,0,0,0,0,0,0,LOT,Rabun County


#### 4. Display the 2 random rows of the DataFrame `df`. Make sure that the rows stay the same if one reruns this cell.

In [5]:
df.sample(2, random_state=0)

,Unnamed: 0,id,stateId,countyId,cityId,country,datePostedString,is_bankOwned,is_forAuction,event,...,parking,garageSpaces,hasGarage,levels,pool,spa,isNewConstruction,hasPetsAllowed,homeType,county
7667,7667,30525-2070209027,16,264690,55543,USA,6/25/2021,0,0,Listing removed,...,0,0,0,0,0,0,0,0,LOT,Rabun County
4197,4197,30530-221908632,16,150983,37982,USA,5/26/2021,0,0,Price change,...,1,2,1,Two,0,0,1,0,SINGLE_FAMILY,Jackson County


#### 5. How many row and columns are in the dataset?

In [6]:
df.shape

(13804, 39)

#### 6. Print list of column names.

In [7]:
df.columns

Index(['Unnamed: 0', 'id', 'stateId', 'countyId', 'cityId', 'country',
       'datePostedString', 'is_bankOwned', 'is_forAuction', 'event', 'time',
       'price', 'pricePerSquareFoot', 'city', 'state', 'yearBuilt',
       'streetAddress', 'zipcode', 'longitude', 'latitude', 'hasBadGeocode',
       'description', 'currency', 'livingArea', 'livingAreaValue',
       'lotAreaUnits', 'bathrooms', 'bedrooms', 'buildingArea', 'parking',
       'garageSpaces', 'hasGarage', 'levels', 'pool', 'spa',
       'isNewConstruction', 'hasPetsAllowed', 'homeType', 'county'],
      dtype='object')

Ensure you understand the meaning of each column! This may require getting hands-on with the data - exploring it thoroughly and even opening the CSV file in Excel for a clearer view.

#### 7. What is `Unnamed: 0` column? Can we delete it without losing essential information? Why?

*Write your answer in this cell. Delete this text and type your answer instead.*

We can delete since it is just a duplicate index column.

The `Unnamed: 0` column typically appears when a dataset is saved with an index in a CSV file, and then read without explicitly setting `index_col=0`. 

#### 8. Delete column `Unnamed: 0`.

In [8]:
df.drop(columns=['Unnamed: 0'], inplace=True, errors='ignore')

#### 9. Display a summary of the basic information for the numerical features in this DataFrame.

In [9]:
pd.options.display.max_columns = None # Use this option to display all columns
df.describe().style.format("{:.1f}")  # I use .style.format("{:.1f}") to make the output more concise

,stateId,countyId,cityId,is_bankOwned,is_forAuction,time,price,pricePerSquareFoot,yearBuilt,zipcode,longitude,latitude,hasBadGeocode,livingArea,livingAreaValue,bathrooms,bedrooms,buildingArea,parking,garageSpaces,hasGarage,pool,spa,isNewConstruction,hasPetsAllowed
count,13804.0,13804.0,13804.0,13804.0,13804.0,13785.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0
mean,16.0,142659.3,32140.4,0.0,0.0,1620448305694.6,367097.3,116.8,1347.4,30670.2,-83.8,33.3,0.0,1858.9,1858.9,1.9,2.3,1177.0,0.5,0.5,0.4,0.1,0.1,0.0,0.0
std,0.0,98056.7,37906.5,0.0,0.0,33739129103.0,647834.6,1748.2,928.1,557.1,1.1,1.1,0.0,43074.4,43074.4,1.9,2.1,1813.7,0.5,0.9,0.5,0.2,0.2,0.2,0.1
min,16.0,17.0,0.0,0.0,0.0,252461000000.0,0.0,0.0,0.0,30002.0,-85.6,30.6,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
25%,16.0,58012.8,13709.0,0.0,0.0,1623280000000.0,95000.0,0.0,0.0,30206.0,-84.5,32.6,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
50%,16.0,116674.5,29242.0,0.0,0.0,1625440000000.0,247000.0,95.0,1965.0,30530.0,-84.1,33.7,0.0,1280.0,1280.0,2.0,3.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
75%,16.0,227609.2,43250.0,0.0,0.0,1625880000000.0,425000.0,161.0,1999.0,31061.0,-83.4,34.0,0.0,2299.2,2299.2,3.0,4.0,1996.2,1.0,0.0,1.0,0.0,0.0,0.0,0.0
max,16.0,333689.0,397383.0,1.0,1.0,1626220000000.0,30504000.0,205000.0,9999.0,31909.0,-80.8,35.0,0.0,5057316.0,5057316.0,89.0,89.0,87120.0,1.0,8.0,1.0,1.0,1.0,1.0,1.0


Review the summary table above and ensure you understand the meaning of each row (e.g., standard deviation, 25th percentile). Then, carefully examine each column to grasp what is happening in the data.

Notice that by default `df.describe()` provides a summary only for *numerical* columns, which is equivalent to `df.select_dtypes(include=['number']).describe()`. To summarize *categorical* columns, use `df.select_dtypes(include=['object', 'category']).describe()`.

Categorical data types in pandas:
- `object` → Often used for text-based categorical data.
- `category` → Optimized for categorical variables, using less memory and enabling faster operations.


#### 10. Display a summary of the basic information for the categorical features in this DataFrame.

In [10]:
df.select_dtypes(include=['object', 'category']).describe()

,id,country,datePostedString,event,city,state,streetAddress,description,currency,lotAreaUnits,levels,homeType,county
count,13804,13804,13803,13785,13804,13804,13804,13705,13804,13804,13804,13804,13804
unique,12546,1,391,5,536,1,12452,12106,1,2,41,5,149
top,31503-93126153,USA,6/25/2021,Listed for sale,Atlanta,GA,Coming Soon Plan,Great Lot in Shadowmoor Subdivision to build y...,USD,Acres,0,SINGLE_FAMILY,Fulton County
freq,2,13804,2103,9520,1031,13804,30,13,13804,9797,6524,8111,1168


Review the summary table above and ensure you understand the meaning of each row (e.g., unique, top, freq). Then, carefully examine each column to grasp what is happening in the data.

#### 11. Print number of unique values in each column of the DataFrame

In [11]:
df.nunique()

id                    12546
stateId                   1
countyId              13697
cityId                  519
country                   1
datePostedString        391
is_bankOwned              2
is_forAuction             2
event                     5
time                    577
price                  2459
pricePerSquareFoot      544
city                    536
state                     1
yearBuilt               168
streetAddress         12452
zipcode                 604
longitude             12250
latitude              12381
hasBadGeocode             1
description           12106
currency                  1
livingArea             3137
livingAreaValue        3137
lotAreaUnits              2
bathrooms                17
bedrooms                 18
buildingArea           2786
parking                   2
garageSpaces              9
hasGarage                 2
levels                   41
pool                      2
spa                       2
isNewConstruction         2
hasPetsAllowed      

#### 12. Some columns contain only a single unique value. Which ones? Print their names. Then, delete all such columns and explain why removing them makes sense.


In [12]:
print(df.columns[df.nunique() == 1])
df = df.loc[:, df.nunique() > 1]

Index(['stateId', 'country', 'state', 'hasBadGeocode', 'currency'], dtype='object')


Columns that contain only one unique value across all rows do not provide any useful information for analysis, because:
- They do not contribute to variability. 
- They do not aid in prediction.
- They take up unnecessary memory.

#### 13. Run the code cell below. Are the results the same or different? If they differ, explain why. 

In [13]:
print((df['livingArea'] == df['livingAreaValue']).all())
print(df['livingArea'].equals(df['livingAreaValue']))

True
True


Answer to the question is *they are the same*.

`(df['col1'] == df['col2']).all()`  
  - Treats `NaN` values as unequal (`NaN != NaN`).  
  - So, if we have missing values, it will report `False`, which is misleading. 

`df['col1'].equals(df['col2'])`  
  - Treats `NaN` values as equal (`NaN == NaN`).  
  - More robust for checking complete equality, including handling missing values. 

#### 14. Display a summary of the numerical features again. Identify any columns with identical statistics. If such columns exist, verify whether they are true duplicates and keep only the first occurrence, removing all others.

In [14]:
df.describe().round(1)

,countyId,cityId,is_bankOwned,is_forAuction,time,price,pricePerSquareFoot,yearBuilt,zipcode,longitude,latitude,livingArea,livingAreaValue,bathrooms,bedrooms,buildingArea,parking,garageSpaces,hasGarage,pool,spa,isNewConstruction,hasPetsAllowed
count,13804.0,13804.0,13804.0,13804.0,1.378500e+04,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0,13804.0
mean,142659.3,32140.4,0.0,0.0,1.620448e+12,367097.3,116.8,1347.4,30670.2,-83.8,33.3,1858.9,1858.9,1.9,2.3,1177.0,0.5,0.5,0.4,0.1,0.1,0.0,0.0
std,98056.7,37906.5,0.0,0.0,3.373913e+10,647834.6,1748.2,928.1,557.1,1.1,1.1,43074.4,43074.4,1.9,2.1,1813.7,0.5,0.9,0.5,0.2,0.2,0.2,0.1
min,17.0,0.0,0.0,0.0,2.524610e+11,0.0,0.0,0.0,30002.0,-85.6,30.6,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
25%,58012.8,13709.0,0.0,0.0,1.623280e+12,95000.0,0.0,0.0,30206.0,-84.5,32.6,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
50%,116674.5,29242.0,0.0,0.0,1.625440e+12,247000.0,95.0,1965.0,30530.0,-84.1,33.7,1280.0,1280.0,2.0,3.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
75%,227609.2,43250.0,0.0,0.0,1.625880e+12,425000.0,161.0,1999.0,31061.0,-83.4,34.0,2299.2,2299.2,3.0,4.0,1996.2,1.0,0.0,1.0,0.0,0.0,0.0,0.0
max,333689.0,397383.0,1.0,1.0,1.626220e+12,30504000.0,205000.0,9999.0,31909.0,-80.8,35.0,5057316.0,5057316.0,89.0,89.0,87120.0,1.0,8.0,1.0,1.0,1.0,1.0,1.0


In [15]:
df = df.drop(columns=['livingAreaValue'])

#### 15. Identify all columns that contain exactly two unique values in the DataFrame and retrieve the unique values present in each of them.

In [16]:
df.loc[:, df.nunique() == 2].apply(pd.unique)

,is_bankOwned,is_forAuction,lotAreaUnits,parking,hasGarage,pool,spa,isNewConstruction,hasPetsAllowed
0,0,0,Acres,0,0,0,0,0,0
1,1,1,sqft,1,1,1,1,1,1


Learn how `.apply()` method works!

#### 16. Something is wrong with `time`. 

The column `time` contains strange numbers. Convert it into a readable datetime format. Keep in mind that the dataset contains property information from 2021.  

Refer to the [Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html) for `to_datetime()` for guidance. Play with the parameter `unit`.

In [17]:
df['time'] = pd.to_datetime(df['time'], unit='ms')
df['time']

0       2021-07-12 00:33:20
1       2021-07-12 00:33:20
2       2021-07-10 01:20:00
3       2021-07-09 00:20:00
4       2021-07-06 00:06:40
                ...        
13799   2021-07-10 01:20:00
13800   2021-07-10 01:20:00
13801   2021-07-09 00:20:00
13802   2021-07-10 01:20:00
13803   2021-07-01 22:53:20
Name: time, Length: 13804, dtype: datetime64[ns]

#### 17. Is there another column contains the same time-related information.

*Write your answer in this cell. Delete this text and type your answer instead.*

It is `datePostedString` column.

#### 18. Let's check for missing values in each column

In [18]:
missing_data = df.isnull().sum().to_frame(name='Missing Values')
missing_data['Percentage'] = (df.isnull().mean() * 100).round(2)
missing_data[missing_data['Missing Values'] > 0]

,Missing Values,Percentage
datePostedString,1,0.01
event,19,0.14
time,19,0.14
description,99,0.72


#### 19. Can the missing values in the `time` column be inferred from other available data? If so, fill in the missing values appropriately.

In [19]:
# Let's check if features 'datePostedString' and 'time' are consistent:

# First let's convert time string to time
df['datePostedString'] = pd.to_datetime(df['datePostedString'], errors='coerce')

print(np.sum(df['time'] != df['datePostedString']))
print(np.sum(df['time'] == df['datePostedString']))


13683
121


They are not always consistent. Let's examine the inconsistencies.

In [20]:
df.loc[df['time'] != df['datePostedString'], ['time','datePostedString']].sample(10)

,time,datePostedString
8613,2021-07-09 00:20:00,2021-07-09
3584,2021-07-01 00:40:00,2021-06-29
12537,2021-07-12 00:33:20,2021-07-09
10247,2021-07-12 22:46:40,2021-07-13
12948,2021-07-10 01:20:00,2021-07-10
5472,2021-05-03 00:00:00,2021-07-10
7622,2021-03-01 23:06:40,2021-06-25
449,2021-06-06 22:53:20,2021-06-07
2514,2021-07-12 00:33:20,2021-07-12
13281,2021-06-04 23:40:00,2021-06-05


The date in the `time` column can be larger, equal, or smaller than `datePostedString`. The difference could be remarkable.

In [21]:
df.loc[(df['time'] - df['datePostedString'] == np.max(df['time'] - df['datePostedString'])), ['time','datePostedString']]

,time,datePostedString
5865,2021-04-12 01:13:20,2014-10-29


In [22]:
df.loc[(df['time'] - df['datePostedString'] == np.min(df['time'] - df['datePostedString'])), ['time','datePostedString']]

,time,datePostedString
5139,1978-01-01 00:03:20,2021-07-06
12319,1978-01-01 00:03:20,2021-07-06


I would say it is NOT advisable to replace missing values in the `time` column with `datePostedString`. However, if we decide to do so, we could use the following commands:

In [23]:
df.loc[df['time'].isnull(), 'time'] = pd.to_datetime(df.loc[df['time'].isnull(), 'datePostedString'])

#### 20. Delete columns `datePostedString` and `description`

In [24]:
df.drop(columns=['datePostedString', 'description'], inplace=True, errors='ignore')

#### 21. Print rows where any value is missing, displaying only the columns: `event`, `time`, and `city`.

In [25]:
df[df.isna().any(axis=1)][['event','time','city']]

,event,time,city
498,NaN,2021-03-25,Atlanta
727,NaN,2021-07-13,Savannah
744,NaN,2021-07-09,Savannah
752,NaN,2021-07-08,Savannah
3982,NaN,2021-06-07,Juliette
4317,NaN,2021-07-09,Guyton
5227,NaN,2021-07-09,Hiawassee
6258,NaN,2021-07-02,Atlanta
7237,NaN,2021-05-20,Thomasville
7340,NaN,2021-06-15,Sandy Springs


#### 22. Change all values `NaN` in the `event` column to `Unknown`

In [26]:
df["event"] = df["event"].fillna("Unknown")

#### 23. Do we still have missing values?

It appears that we no longer have any explicit missing values (NaN). However, missing data can sometimes be represented by zeros or small values. Identify all columns where this might be the case.

In [27]:
suspicious_columns = df.columns[((df == 0).sum() > 0) & (df.nunique() > 2)]
suspicious_columns

Index(['cityId', 'price', 'pricePerSquareFoot', 'yearBuilt', 'livingArea',
       'bathrooms', 'bedrooms', 'buildingArea', 'garageSpaces'],
      dtype='object')

#### 24. In all those columns replace zeros with `NaN`s

In [28]:
df[suspicious_columns] = df[suspicious_columns].replace(0, np.nan)

#### Can we reliably infer missing `cityId` values using `city`?

To answer such question, we need to understand a few things.

#### 25. Do all cities have unique `cityId`? Find cities wuth multiple IDs if any.

In [29]:
unique_city_counts = df.groupby('city')['cityId'].nunique() # .nunique() counts only non-null unique values by default
unique_city_counts[unique_city_counts > 1]

city
Acworth         4
Adairsville     2
Adel            2
Ailey           2
Alamo           3
               ..
Woodbine        2
Woodstock       2
Wrens           2
Wrightsville    2
Zebulon         3
Name: cityId, Length: 237, dtype: int64

#### 26. Does every city have at least one `cityId`? Identify any cities without a `cityId` if they exist.

In [30]:
unique_city_counts[unique_city_counts  < 1]

city
Box Springs      0
Byromville       0
Harrison         0
Jekyll Island    0
Newington        0
Norman park      0
Rabun Gap        0
Register         0
Richland         0
Stephens         0
Tate             0
Waco             0
Name: cityId, dtype: int64

#### 27. Replace `NaN` values in the `cityId` column with the corresponding `cityId` for cities that have a unique `cityId`. Report how many missing values where in the `cityId` column before and after the replacement.

In [31]:
print(f'before {df['cityId'].isna().sum()}')

cities_with_one_id = unique_city_counts[unique_city_counts == 1].index
city_to_cityId = df[df['city'].isin(cities_with_one_id)].groupby('city')['cityId'].max()

df['cityId'] = df['cityId'].fillna(df['city'].map(city_to_cityId))

print(f'after {df['cityId'].isna().sum()}')

before 911
after 689


#### Whew! That was a challenge, but now it's time for a few more tweaks and discoveries.

#### 28. Identify all columns that contain exactly two distinct values and convert them to categorical.

In [32]:
binary_cols = df.columns[df.nunique() == 2]
df[binary_cols] = df[binary_cols].astype('category')

#### 29. It makes sense to convert the columns `countyId`, `cityId`, and `zipcode` into categorical. Apply the conversion.	

In [33]:
df[['countyId', 'cityId', 'zipcode']] = df[['countyId', 'cityId', 'zipcode']].astype('category')

#### 30. Determine if there are any duplicate rows in the DataFrame. If duplicates exist, remove them, keeping only the first occurrence.

In [34]:
df[df.duplicated()]

,id,countyId,cityId,is_bankOwned,is_forAuction,event,time,price,pricePerSquareFoot,city,yearBuilt,streetAddress,zipcode,longitude,latitude,livingArea,lotAreaUnits,bathrooms,bedrooms,buildingArea,parking,garageSpaces,hasGarage,levels,pool,spa,isNewConstruction,hasPetsAllowed,homeType,county


There are no duplicate records in the table.

#### 31. Display a summary of the basic information for numerical, categorical and object features in this DataFrame.

In [35]:
pd.options.display.max_columns = None # Use this option to display all columns
df.describe().style.format("{:.1f}")

,time,price,pricePerSquareFoot,yearBuilt,longitude,latitude,livingArea,bathrooms,bedrooms,buildingArea,garageSpaces
count,13804.0,13755.0,8372.0,9381.0,13804.0,13804.0,8407.0,9107.0,9122.0,6445.0,3262.0
mean,.1f,368405.0,192.6,1982.7,-83.8,33.3,3052.2,2.8,3.6,2520.9,2.1
min,.1f,10.0,2.0,1790.0,-85.6,30.6,1.0,1.0,1.0,1.0,1.0
25%,.1f,98000.0,109.0,1963.0,-84.5,32.6,1440.0,2.0,3.0,1465.0,2.0
50%,.1f,249000.0,146.0,1989.0,-84.1,33.7,2030.0,3.0,3.0,2099.0,2.0
75%,.1f,425000.0,198.0,2005.0,-83.4,34.0,2930.0,3.0,4.0,3026.0,2.0
max,.1f,30504000.0,205000.0,9999.0,-80.8,35.0,5057316.0,89.0,89.0,87120.0,8.0
std,nan,648616.3,2241.6,88.4,1.1,1.1,55163.5,1.6,1.5,1912.6,0.6


In [36]:
df.select_dtypes(include=['category']).describe()

,countyId,cityId,is_bankOwned,is_forAuction,zipcode,lotAreaUnits,parking,hasGarage,pool,spa,isNewConstruction,hasPetsAllowed
count,13804,13115.0,13804,13804,13804,13804,13804,13804,13804,13804,13804,13804
unique,13697,518.0,2,2,604,2,2,2,2,2,2,2
top,84489,37211.0,0,0,30165,Acres,0,0,0,0,0,0
freq,2,731.0,13802,13798,78,9797,7450,8906,13053,12966,13172,13751


In [37]:
df.select_dtypes(include=['object']).describe()

,id,event,city,streetAddress,levels,homeType,county
count,13804,13804,13804,13804,13804,13804,13804
unique,12546,6,536,12452,41,5,149
top,31503-93126153,Listed for sale,Atlanta,Coming Soon Plan,0,SINGLE_FAMILY,Fulton County
freq,2,9520,1031,30,6524,8111,1168


#### 32.  What interesting findings can you get from these tables and what question can you ask? Name two.

Below are some examples:

**Finding 1:**  The binary feature `is_bankOwned` has only two values:
- 0 (Not Bank-Owned): 13,802 properties  
- 1 (Bank-Owned): Only 2 properties  

**Question:** Why is this the case? Is it due to missing data, dataset bias, or other factors?  

---

**Finding 2:**  The feature `pricePerSquareFoot` has an extreme outlier $205,000 per square foot.  
- The total price of this property is also $205,000, which strongly suggests a data entry error.  

**Question:** Are there any other similar errors in the dataset?


*Write your answer here.*